### ISSUE TITLE
[Ray Dataset|Ray AIR] Streaming Executor is disabled in TensorflowTrainer train loop  

### ISSUE THREAD

**Issue by hanli-0612** [ray-project/ray#32586]
### What happened + What you expected to happen

I'm working on a test script with Ray nightly release to try out streaming executor. I found the streaming executable dataset turns back to bulkly executor after wrapping Ray AIR trainer around the dataset.

Here's the code snippet on which streaming executor runs well:
```
import ray
import time
ray.data.context.DatasetContext.get_current().use_streaming_executor = True

def sleep(x):
    time.sleep(0.1)
    return x

for _ in (
    ray.data.range_tensor(5000, shape=(80, 80, 3), parallelism=200)
    .map(sleep)
    .map_batches(sleep, num_cpus=1)
    .map_batches(lambda x: x, batch_format="numpy")
    .iter_batches()
):
    pass
```

Here's the code snippet with Tensorflow Trainer introduced, and executor is set back to bulky executor:
```
import ray
import time
from ray.air import session
from ray.train.tensorflow import TensorflowTrainer
from ray.air.config import FailureConfig, RunConfig, ScalingConfig, DatasetConfig

ray.data.context.DatasetContext.get_current().use_streaming_executor = True

def sleep(x):
    time.sleep(0.1)
    return x

def debug(config):
    train_dataset_pipeline_shard = session.get_dataset_shard("train")
    for i in range(20):
        for batch in train_dataset_pipeline_shard.iter_batches():
            pass
        print(f"Epoch {i}")

dataset = ray.data.range_tensor(5000, shape=(80, 80, 3), parallelism=200).map(sleep).map_batches(sleep, num_cpus=1).map_batches(lambda x: x, batch_format="numpy")

trainer = TensorflowTrainer(
    train_loop_per_worker=debug,
    run_config=RunConfig(
        failure_config=FailureConfig(max_failures=5),
    ),
    datasets={
        "train": dataset,
    },
    train_loop_config={},
    dataset_config={
        "train": DatasetConfig(max_object_store_memory_fraction=0.2),
    },
    scaling_config=ScalingConfig(num_workers=2, use_gpu=True),
)

trainer.fit()
```
Logs out of this run: "bulk_executor.py:39 -- Executing DAG InputDataBuffer[Input] -> AllToAllOperator[randomize_block_order]"

### Versions / Dependencies

nightly build as of 02/14/2023

### Reproduction script

provided in description

### Issue Severity

High: It blocks me from completing my task.

**Comment by c21** [comment]
cc @amogkam and @jianoaix.

**Comment by matthewdeng** [comment]
Hmm is it because `randomize_block_order` is an `AllToAllOperator`?

**Comment by hanli-0612** [comment]
is it because AIR trainer shuffles the block?

**Comment by clarkzinzow** [comment]
@c21 @matthewdeng @amogkam @jianoaix Isn't this simply because `max_object_store_memory_fraction` being set will put AIR down the `DatasetPipeline` path, which (currently) always uses the bulk executor? https://github.com/ray-project/ray/blob/a3beb655f838d6e07ce888e1d706f2d91beb5a93/python/ray/train/_internal/dataset_spec.py#L185-L210

**Comment by jianoaix** [comment]
Yeah, `randomize_block_order` should work.

**Comment by clarkzinzow** [comment]
Also, even if `max_object_store_memory_fraction` wasn't set and we weren't using `DatasetPipeline`s, wouldn't the underlying dataset split trigger bulk execution since `num_workers=2` suggests data parallel training?

**Comment by amogkam** [comment]
Yes that's right as well @clarkzinzow

**Comment by hanli-0612** [comment]
FYI after comment out I still see bulk execution:
```
INFO bulk_executor.py:39 -- Executing DAG InputDataBuffer[Input] -> TaskPoolMapOperator[read->map->MapBatches(sleep)->MapBatches(<lambda>)] -> AllToAllOperator[randomize_block_order]
```

**Comment by amogkam** [comment]
Yeah that's right @hanli-0612...streaming executor is currently not supported with AIR Trainers right now.

**Comment by hanli-0612** [comment]
Thanks @amogkam for confirming this. Will it be something to be supported in the future? It'll be super useful feature to train large dataset in distributed way.

**Comment by amogkam** [comment]
Yes, this absolutely will be supported in the future. Unfortunately, at present it does not work with AIR Trainers.

**Comment by ericl** [comment]
Yup, we will be adding this soon.

In the mean-time, you can try out streaming ingest by creating a Dataset directly from your training process, e.g., example 2 in the dev preview guide here: https://docs.google.com/document/d/1BXd1cGexDnqHAIVoxTnV3BV0sklO9UXqPwSdHukExhY/edit#heading=h.dkdm0ffungpx

This will be dropping some of the AIR niceties, such as automatic preprocessor fitting, but can be used to try out the new path.

**Comment by vsriram23** [comment]
Hi, can AIR trainer use streaming execution now ?
Is there any new release that can be used to test this functionality ?

**Comment by ericl** [comment]
Hi, a quick update on this issue:

1. We've recently merged support for streaming split, and this is available in nightly wheels and also will be in the upcoming 2.4 release. This feature is finalized, and you can find the documentation here: https://docs.ray.io/en/master/data/api/doc/ray.data.Dataset.streaming_split.html

2. The final design of integrating streaming Datasets execution with Ray Train is still under discussion here: https://github.com/ray-project/enhancements/pull/26, and is currently targeted to land in 2.5.

Right now, if you'd like to try out using streaming mode for distributed training, I would recommend using the streaming split API directly, bypassing the "datasets" arg in Train. It would look something like this:

```
ds = ray.data.read(...)
splits: List[DatasetIterator] = ds.streaming_split(NUM_WORKERS, equal=True)

def my_train_func(...)
    rank = session.get_local_rank()
    split = splits[rank]
    for _ in range(NUM_EPOCHS):
        for batch in split.iter_batches():
            ...

trainer = TorchTrainer(my_train_func)
trainer.fit()

```

**Comment by vsriram23** [comment]
Hi, 

The provided link for documentation seems to be not working
https://docs.ray.io/en/master/data/api/doc/ray.data.Dataset.streaming_split.html

Can anyone provide a new link ?

Is the split usage straight-forward with Ray-tune similar to Trainer as shown above ?

**Comment by ericl** [comment]
You can find the streaming split docs here: https://docs.ray.io/en/master/data/api/doc/ray.data.Datastream.streaming_split.html

> Is the split usage straight-forward with Ray-tune similar to Trainer as shown above ?

It will be a little trickier with Ray Tune until the Train<>Tune API is updated. You'll need to call split in each trial's main actor and pass handles to sub actors separately, since you can't have all trials sharing the same stream.

**Comment by vsriram23** [comment]
> > Is the split usage straight-forward with Ray-tune similar to Trainer as shown above ?
> 
> It will be a little trickier with Ray Tune until the Train<>Tune API is updated. You'll need to call split in each trial's main actor and pass handles to sub actors separately, since you can't have all trials sharing the same stream.

Can you show a sample code snippet on how to do this ? Thanks.

**Comment by ericl** [comment]
Yup, you can do global shuffles per-epoch as follows:

```
ds = ds.random_shuffle()   # This will be re-evaluated every epoch.

Trainer(
   datasets={"train": ds}
)
```

**Comment by vsriram23** [comment]
@ericl , just to confirm will the below code also update splits for each epoch ? Thanks.

ds = ray.data.read(...)
ds = ds.random_shuffle()
splits: List[DatasetIterator] = ds.streaming_split(NUM_WORKERS, equal=True)

def my_train_func(...)
    rank = session.get_local_rank()
    split = splits[rank]
    for _ in range(NUM_EPOCHS):
        for batch in split.iter_batches():
            ...

trainer = TorchTrainer(my_train_func, datasets={"train": ds})
trainer.fit()

**Comment by ericl** [comment]
That's right. The full shuffle is a blocking operation and runs on each epoch before data is sent to splits.
